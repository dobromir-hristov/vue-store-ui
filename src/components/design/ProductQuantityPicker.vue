<template>
  <div class="ProductQuantityPicker">
    <label class="mb-2 block">Quantity</label>
    <div>
      <input
        v-model.number="quantity"
        type="number"
        class="w-16 border-b border-gray pb-2"
      >
    </div>
  </div>
</template>

<script>
import { ProductQuantityMixin } from '@/mixins/core/ProductQuantityMixin'

/**
 * Renders a quantity picker for the current product
 * @module ProductQuantityPicker
 */
export default {
  name: 'ProductQuantityPicker',
  mixins: [ProductQuantityMixin]
}
</script>
