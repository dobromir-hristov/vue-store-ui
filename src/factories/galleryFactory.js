export const galleryFactory = () => ([{
  variation_id: 'color__blue',
  src: 'https://demo.storefrontcloud.io/img/600/744/resize/w/s/ws12-blue_main.jpg'
}, {
  variation_id: 'color__orange',
  src: 'https://demo.storefrontcloud.io/img/600/744/resize/w/s/ws12-orange_main.jpg'
}, {
  variation_id: 'color__purple',
  src: 'https://demo.storefrontcloud.io/img/600/744/resize/w/s/ws12-purple_main.jpg'
}])
