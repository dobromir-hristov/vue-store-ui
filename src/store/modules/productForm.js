import { ProductForm } from '@/models/ProductForm'

const PREPARE_PRODUCT_FORM = 'PREPARE_PRODUCT_FORM'

const state = {
  /** @type ProductForm */
  productForm: null
}

const getters = {
  productForm: state => state.productForm,
  /**
   * Returns the SKU of a product based on the selected variations
   * @param state
   * @return {string}
   */
  formSKU (state) {
    if (!state.productForm) return ''
    return state.productForm.sku + '-' + Object.values(state.productForm.variations).join('-')
  }
}

const mutations = {
  [PREPARE_PRODUCT_FORM] (state, product) {
    state.productForm = new ProductForm(product)
  }
}

const actions = {
  prepareForm ({ commit }, product) {
    commit(PREPARE_PRODUCT_FORM, product)
  }
}

export default {
  state,
  getters,
  mutations,
  actions,
  namespaced: true
}
