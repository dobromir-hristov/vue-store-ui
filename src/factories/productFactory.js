import { galleryFactory } from '@/factories/galleryFactory'

export const productFactory = () => ({
  id: 'some-random-id',
  name: 'Radiant Tee',
  sku: 'radiant-tee',
  price: 22.50,
  currency: 'USD',
  description: `<p>So light and comfy, you'll love the Radiant Tee's organic fabric, feel, performance and style. You may never want to stop moving in this shirt.</p>
<p>• Salmon soft scoop neck tee.<br>• Athletic, semi-form fit.<br>• Flat seams prevent chafing.<br>• 67% Organic Cotton / 28% Hemp / 5% Spandex.</p>`,
  attributes: [
    { type: 'material', values: ['Hemp', 'Organic Cotton', 'Spandex'] },
    { type: 'style', values: ['Tee'] },
    { type: 'climate', values: ['Indoor', 'Warm'] },
    { type: 'pattern', values: ['Solid'] }
  ],
  variations: [{
    id: 'color',
    type: 'color',
    name: 'Color',
    price: 5,
    types: [
      { id: 'blue', name: 'Blue', value: 'blue' },
      { id: 'purple', name: 'Purple', value: 'purple' },
      { id: 'orange', name: 'Orange', value: 'orange' }
    ]
  }, {
    id: 'size',
    type: 'size',
    name: 'Size',
    price: 10,
    types: [
      { id: 'm', name: 'Medium', value: 'm' },
      { id: 'l', name: 'Large', value: 'l' },
      { id: 'xl', name: 'XLarge', value: 'xl' }
    ]
  }, {
    id: 'material',
    type: 'material',
    name: 'Material',
    price: 0,
    types: [
      { id: 'cotton', name: 'Cotton', value: 'cotton' },
      { id: 'cot50poli50', name: 'Cotton 50%, Polyester 50%', value: 'cot50poli50' }
    ]
  }
  ],
  categories: [
    { id: '2', name: 'Default Category' },
    { id: '3', name: 'Women' },
    { id: '4', name: 'Tops' },
    { id: '5', name: 'Tees' }
  ],
  gallery: galleryFactory()
})
