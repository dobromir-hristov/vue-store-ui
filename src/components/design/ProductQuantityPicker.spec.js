import ProductQuantityPicker from './ProductQuantityPicker'
import productForm from '@/store/modules/productForm'
import { productFactory } from '@/factories/productFactory'

/**
 * @param overrides
 * @returns {{config, wrapper: Wrapper<V> | Wrapper<Vue> | *}}
 */
const createWrapper = async (overrides) => {
  const store = {
    'product/productForm': productForm
  }

  let config = _merge(createComponentMocks({ store }), overrides)
  await config.store.dispatch('product/productForm/prepareForm', productFactory())
  const wrapper = mount(ProductQuantityPicker, config)
  return {
    config,
    wrapper
  }
}

describe('ProductQuantityPicker', () => {
  it('renders the ProductQuantityPicker', async () => {
    const { wrapper } = await createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('increases the quantity in the store from the picker', async () => {
    const { wrapper } = await createWrapper()
    wrapper.find('input').setValue(5)
    expect(wrapper.vm.$store.getters['product/productForm/quantity']).toBe(5)
  })
})
