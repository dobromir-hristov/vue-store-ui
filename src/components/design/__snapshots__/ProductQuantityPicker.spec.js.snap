// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductQuantityPicker renders the ProductQuantityPicker 1`] = `
<div class="ProductQuantityPicker"><label class="mb-2 block">Quantity</label>
  <div><input type="number" class="w-16 border-b border-gray pb-2"></div>
</div>
`;
